proto/proxy/destination/destination.proto

```diff
@@ -10,12 +10,40 @@ import "common/common.proto";
 // weighted set of addresses and address metadata.  Can be implemented with DNS
 // or lookups against other service discovery backends.
 //
+// If the service does not exist then the controller must send
+// `no_endpoints{exists: false}` ASAP when a client subscribes or when the
+// service stops existing. If the service exists and has endpoints available
+// then the controller must send `add` that lists all (or at least a large
+// number) of the endpoints for the service. If and only if the service exists
+// but does not have any endpoints available then the controller SHOULD send
+// `no_endpoints{exists: true}` when a client subscribes. In other words, the
+// `no_endpoints` message must only be sent when there are *no*endpoints for
+// the service.
+//
 // The controller is expected to send an Update every time there is a
 // change in service discovery.  The controller is also expected to send an
 // update at least once every ADDRESS_UPDATE_INTERVAL to indicate that the
 // controller is still healthy.  If no service discovery updates have taken
 // place, the controller can simply send an empty `add`.  The controller may
 // determine the value of ADDRESS_UPDATE_INTERVAL.
+//
+// The client MUST be prepared to receive messages in any order and the client
+// MUST be able to cope with the presence or absence of redundant messages.
+//
+// `no_endpoints` followed by an `add` is *not* equivalent to just sending the
```

Comment:
`[NoEndpoints(true); Add(1.1.1.1)]` will first clear the client's set of addresses and then add 1.1.1.1 to it. `[Add(1.1.1.1)]` will add 1.1.1.1 to the client's set and then remove all other entries. So, the former results in a temporary state where there are no addresses in the client's set. The latter results in there not being a temporary state where there are no addresses.

Reply:
Gotcha.  I think this comment is confusing because the two sequences of messages are equivalent in that they both end up with the client in the same state (but go through different intermediate states).  I think it safely goes without saying that `[NoEndpoints(true)]` is not equivalent to `[Add(1.1.1.1)]`.